### PR TITLE
confgenerator: add stringer implementation for UnifiedConfig

### PR DESCRIPTION
**Pull request by ridwanmsharif**
[b/235585629](http://b/235585629) | P3 | Confgenerator serializes and de-serializes the config unnecessarily

This change does the following:
- Add stringer implementation to the unified config
- Removes the need for the extraneous serialization and deserialization
- Cleans up some dead code
- Makes it so that an empty list for include_paths is a valid use case

### CODE REVIEW

**Comment by sophieyfang** [comment]
Centos 7 regex shall be activemq:
image

File: confgenerator/config.go
Comment (braydonk) [inline confgenerator/config.go:70]:
```suggestion
		return fmt.Sprintf("failed to convert Unified config to yaml: %v", err)
```

File: confgenerator/files.go
Comment (braydonk) [inline confgenerator/files.go:38]: Doesn't change anything functionally, but conventionally statements like this will have an `err` just for this scope.
```suggestion
	if err := uc.Validate(platform); err != nil {
```

File: confgenerator/confgenerator_test.go
Comment (braydonk) [inline confgenerator/confgenerator_test.go:186]: Every time `Validate` is called, or really any of the `UnifiedConfig` methods, we have to constantly pass in the platform. Maybe it should just be a private field?

Also maybe an aside, but while looking at this I noticed `uc.Logging.Validate` takes in a platform but doesn't use it. Also, I don't know why we have a `uc.ValidateMetrics` but also `uc.Logging.Validate` I don't understand why that API is different. Might be out of scope for this PR though.

File: confgenerator/confgenerator_test.go
Comment (braydonk) [inline confgenerator/confgenerator_test.go:184]: Maybe built in config could be a field on the Unified Config instead of always needing to look up from `apps`? That or `apps` could have a public function that gets the built in configuration for that platform.

File: confgenerator/testdata/invalid/linux/logging-receiver_files_type_invalid_parameter_include_paths_is_empty_list/golden/error
Comment (braydonk) [inline confgenerator/testdata/invalid/linux/logging-receiver_files_type_invalid_parameter_include_paths_is_empty_list/golden/error:1]: Why is this test valid now?

File: cmd/ops_agent_windows/run_windows.go
Comment (jefferbrecht) [inline cmd/ops_agent_windows/run_windows.go:134]: This is a good opportunity to refactor these magic numbers out to a named constant.

File: cmd/google_cloud_ops_agent_diagnostics/main.go
Comment (ridwanmsharif) [reply cmd/google_cloud_ops_agent_diagnostics/main.go:30]: So I actually think the behaviour now is strictly better than before. I didn't see any good reason why the empty list for `include_paths` was ever prohibited, it doesn't break anything. Our validation on the field says it is `required` to be valid, not that it must be non-empty. If we do want that as well, we should add that as a validator and not rely on marshalling with `omitempty` and then unmarshalling again to validate this one validation rule.

Also the behaviour before was kinda broken. It would surface errors on the merged config and give yaml line numbers and show the error on the merged config. Which is never visible to the users anyway, so the errors were less than ideal too. The merged config should be an internal thing, and users should only interact with the user specified config.

Let me know if you think I should add a validator to ensure the old behaviour regarding us requiring non-empty `include_paths` (which we should also document) or whether we should loosed the requirements as we do here.

Comment (jefferbrecht) [reply cmd/google_cloud_ops_agent_diagnostics/main.go:30]: YAML and validation are orthogonal. That we inline validation with unmarshalling is a convenience and not a requirement: it is absolutely possible to run struct validation entirely on its own without having to marshal/unmarshal anything to/from YAML.

The marshal/unmarshal round-trip should be a no-op. YAML is only an intermediate representation and should be state-preserving. The YAML tags for `include_paths` allow valid things to become invalid after a round-trip test, therefore `omitempty` was the wrong tag to use. If the original intent was that `include_paths` must not be empty then that should have been encoded entirely within the `validate` tag, e.g. `min=1` instead of `required`.

What is the use case for `include_paths` being empty? "That it doesn't break things" isn't the bar we should meet; the bar should at least be "we don't want the user to do the wrong thing", and an empty `include_paths` is clearly the wrong thing IMO, and should be disallowed.

Regarding where to put the struct-level validation: I would go the other way and call `uc.Validate` from within `MergeConfFiles`, which has the additional benefit of simplifying all of the call sites. I can't think of a reason why the intermediate state after-merge-but-before-validate would be useful so it should be OK to combine those steps.

For the missing struct-level validation, I would still be OK with assuming that valid inputs always produce a valid output during `MergeConfFiles`, which implies that struct-level validation should be performed inside there and issue a `panic` on failure.

Comment (jefferbrecht) [reply cmd/google_cloud_ops_agent_diagnostics/main.go]: Looks good 👍 We might also be able to get away with moving `original.Validate` to the panic as well, since `original` is just the built-in struct rather than something that's user-provided.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (braydonk) [inline cmd/google_cloud_ops_agent_engine/main.go:50]: Sorry, I was wrong; `fmt.Printf` doesn't add the new lines but the `log` package actually prints everything on a new line regardless. Sorry!

File: confgenerator/files.go
Comment (braydonk) [inline confgenerator/files.go:34]: this else isn't necessary

File: cmd/google_cloud_ops_agent_diagnostics/main.go
Comment (braydonk) [inline cmd/google_cloud_ops_agent_diagnostics/main.go:37]: here as well

File: confgenerator/files.go
Comment (braydonk) [inline confgenerator/files.go:41]: If you just return nil here then you don't need to instantiate on line 38

File: confgenerator/files.go
Comment (braydonk) [inline confgenerator/files.go:43]: UnmarshalYamlToUnifiedConfig really should return a reference, should we do that in this PR?

Comment (ridwanmsharif) [reply confgenerator/files.go:43]: Might as well. Done.

**Comment by ridwanmsharif** [comment]
I wonder if I'll ever see any of my PRs have only green checks. Or will the windows and sles gods always cast a shadow on me. Investigating the failures now. Hoping they are unrelated.

**Comment by ridwanmsharif** [comment]
Okay the test failures do appear to be unrelated and flaky. Don't think they are blocking the PR.

**Review by braydonk — APPROVED** [review]

**Pull request merged**
